src/compiler_lectures/how_salsa_works.md

```diff
@@ -0,0 +1,169 @@
+# How Salsa works
+
+This chapter is based on the explanation given by Niko Matsakis in this [video](https://www.youtube.com/watch?v=_muY4HjSqVw) about [Salsa](https://github.com/salsa-rs/salsa).
+
+## What is Salsa?
+
+Salsa is a library for incremental recomputation, this means reusing computation that has already been done in the past to increase the efficiency of future computations.
+
+The objectives of Salsa are:
+ * Provide that functionality in an automatic way, so reusing old computations is done automatically by the library
+ * Doing so in a "sound", or "correct", way, therefore leading the same results as if it had been done from scratch
+
+Salsa actual model is much richer, allowing many kinds of inputs and many different outputs.
+For example, integrating Salsa with an IDE could mean that the inputs could be the manifest (`Cargo.toml`), entire source files (`foo.rs`), snippets and so on; the outputs of such integration could range from a binary executable, to lints, types (for example, if a user selects a certain variable and wishes to see it's type), completitions etcetera.
+
+## How does it work?
+
+The first thing that Salsa has to do is identify the "base inputs" [^EN1].
+
+Then Salsa has to also identify intermidiate, "derived" values, which are something that the library produces, but, for each derived value there's a "pure" function that computes the derived value.
+
+For example, there might be a function `ast(x: Path) -> AST`. The produced `AST` isn't a final value, it's an intermidiate value that the library would use for the computation.
+
+This means that when you try to compute with the library, Salsa is going to compute various derived values, and eventually read the input and produce the result for the asked computation.
+
+Salsa is going to track, in the course of computing, which inputs were accessed, which derived values, and this is going to be used later to determine what's going to happen when the inputs change: are the derived values still valid? 
+
+This doesn't mean necessarily that each computation downstream from the input is going to be checked, since that is going to be costly. Salsa only needs to check each downstream computation until it finds one that isn't changed, therefore it won't need to check the other derived computations, since they wouldn't need to change.
```

Comment:
```suggestion
This doesn't necessarily mean that each computation downstream from the input is going to be checked, which could be costly. Salsa only needs to check each downstream computation until it finds one that isn't changed. At that point, it won't check other derived computations since they wouldn't need to change.
```